### DISCUSSION TITLE
Define fields that are not arguments to the constructor

### DISCUSSION THREAD

**Discussion by felix-hilden** [pydantic/pydantic#5929] [Question] [answered]
I'd like to define a field that is not available in the model arguments. So to define an instance attribute with either a default or a default factory, and no way to change it when constructing the model (assigning is fine). To essentially achieve this:

```python
class Foo:
    def __init__(self, a: int, b: str = ""):
        self.a = a
        self.b = b
        self.c = generate_c()  # or without the call as just a default value
```

Does Pydantic support this? I didn't find relevant arguments to `Field` that would do it, or any information when searching online. I see that private (ignored) fields are available, but I'd like for this to be compatible with validation and inheritance. So I could write for example:

```python
class Foo(BaseModel):
    a: int
    b: str = ""
    c: int

class Bar(Foo):
    c: int = Field(default_factory=list, no_init=True)
```

**Comment by adriangb** [comment]
We do not currently support this, but dataclasses do (via `init=False`) and it's a reasonable feature request for Pydantic.

**Reply by felix-hilden** [reply]
Thank you for the quick response! May I create an issue (or can this discussion be converted to one)?

**Reply by felix-hilden** [reply, marked as answer]
The support has been added in 2.6 it seems!

**Comment by lig** [comment]
Also, `init` specified in PEP 683 as a standard field parameter https://peps.python.org/pep-0681/#field-specifier-parameters.

**Comment by guifry** [comment]
I would also like `init` to be added to Pydantic as a boolean `Field` parameter.

**Comment by lig** [comment]
Pydantic uses `Field(init_var: bool)` for that. See https://docs.pydantic.dev/latest/usage/fields/#dataclass-constraints

**Reply by felix-hilden** [reply]
That's a very useful arg as well, but not quite what I was after. The new `init` does the trick though 🎉

**Reply by lautig32** [reply]
I'm in the same situation, I'm trying to make the attribute that I set the variable to not be considered in the initialization even if it's sent as a parameter, but I can't get it to work... Do you know if this possibility exists?

**Comment by batzkass** [comment]
I don't understand why the feature behind this question was considered as supported since 2.6. The documentation [here](https://docs.pydantic.dev/latest/usage/fields/#dataclass-constraints) concerns the case of `dataclass` usage, not Pydantic `BaseModel`.

I may have missed something, tell me if I should post a feature request.

In the case of BaseModel, using Field's `init=True` argument seems to have no effect. In my use case, exactly as @felix-hilden initially asked, I need some fields that can't be set on model instantiation, but assignation after that is okay. Private attrs are not an option for two reason: I can't prefix all them with "_", and I want some of them to appear in serialization (hence not really private).

To workaround this, for now I subclassed BaseModel:

```python
from pydantic import BaseModel

class MyBaseModel(BaseModel):
    def __init__(self, **kwargs):
        for kwarg_name in list(kwargs.keys()):
            if(kwarg_name in self.__fields__ and self.__fields__[kwarg_name].init==False):
                kwargs.pop(kwarg_name, None)
        return super().__init__(**kwargs)
```
I didn't tested that extensively, but this way I can use `Field([...], init=True)` just like in dataclasses case, and it does what requested.

**Reply by ilanbiala** [reply]
Same, this also seems like a miss to me.

**Reply by leclairm** [reply]
Agreed.
